### PR TITLE
Comment History

### PR DESCRIPTION
**Pull request by zRyder**
In Short, comment history requesting and response parsing works and better fits the idea behind dash-rs. Hi Stadust please merge me

### CODE REVIEW

**Comment by zRyder** [comment]
Totally gonna clean this up later maybe probably

File: src/response.rs [hunk 1]
```diff
@@ -193,3 +227,16 @@ pub fn parse_get_gj_acccount_comments_response(response: &str) -> Result<Vec<Pro
...
+    use crate::response::parse_get_gj_comment_history_response;
+
+    #[test]
+    fn test_parse_comment_history_request() {
+        let test = parse_get_gj_comment_history_response("2~Rmlyc3Q=~1~76298358~3~3713125~4~2~10~0~9~1 week~6~33134786~11~2~12~75,255,75:1~Ryder~9~101~10~7~11~9~14~0~15~2~16~57903|2~TG92ZSB0byBzZWUgaXQ=~1~69201939~3~3713125~4~5~10~0~9~2 weeks~6~32301881~11~2~12~75,255,75:1~Ryder~9~101~10~7~11~9~14~0~15~2~16~57903#999:0:2");
+
+        println!("{:?}", test);
+    }
+
+}
```
Comment (qimiko) [inline src/response.rs:242]: i don't think the tests are supposed to be like this, aren't they supposed to be in the tests folder and also... check that the function returned what it's supposed to

Comment (zRyder) [reply src/response.rs]: Testing was there solely for me, typically we don't test on the response parsing functions, removed.

**Comment by qimiko** [comment]
talked about this a little bit in the discord but to keep a note of it:

with gd's internals, comment history and comments both use the same object/parser. this pull request creates a new model for history comments, which doesn't match gd exactly. considering partial levels and level/profile comments exist, this might fit in line with what the rest of dash-rs does?

File: src/model/comment/level.rs [hunk 2]
```diff
@@ -76,6 +75,67 @@ pub struct LevelComment<'a> {
     pub special_color: Option<Color>,
 }
 
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+pub struct CommentHistory<'a> {
+    /// Information about the user that made this [`LevelComment`]. Is generally a [`CommentUser`]
```
Comment (stadust) [inline src/model/comment/level.rs:80]: drop the "generally"

File: src/model/mod.rs [hunk 3]
```diff
@@ -68,3 +69,7 @@ impl Display for GameVersion {
         }
     }
 }
+
+pub trait FromRobTopStr {
+    fn from_robtop_str(robtop_str: &str) -> Result<Self, DeError> where Self: Sized;
+}
```
Comment (stadust) [inline src/model/mod.rs:75]: what's this trait for?

File: src/model/user/profile.rs [hunk 4]
```diff
@@ -1,9 +1,6 @@
 use crate::model::user::{Color, ModLevel};
-use serde::{Deserialize, Serialize};
-use std::{
-    borrow::Cow,
-    fmt::{Display, Formatter},
-};
+use serde::{__private::Formatter, Deserialize, Serialize};
```
Comment (stadust) [inline src/model/user/profile.rs:2]: wrong import?

File: src/response.rs [hunk 5]
```diff
@@ -14,7 +14,9 @@ use crate::{
     },
     DeError, HasRobtopFormat,
 };
-use std::fmt::{Display, Formatter};
+
+use serde::__private::Formatter;
```
Comment (stadust) [inline src/response.rs:18]: wrong import

File: src/response.rs [hunk 6]
```diff
@@ -181,6 +183,38 @@ pub fn parse_get_gj_comments_response(response: &str) -> Result<Vec<LevelComment
...
+                comment.user = if raw_user == "1~~9~~10~~11~~14~~15~~16~" {
+                    None
+                } else {
+                    Some(CommentUser::from_robtop_str(raw_user)?)
+                };
+
+                Ok(comment)
+            } else {
+                Err(ResponseError::UnexpectedFormat)
+            }
```
Comment (stadust) [inline src/response.rs:213]: iirc this is copy pasted from the level comment processes so maybe there's a way to unify them

File: src/serde/de/error.rs [hunk 7]
```diff
@@ -1,7 +1,7 @@
 //! Module containing the error type for deserialization errors
 
-use std::fmt::{Display, Formatter};
-
+use serde::__private::Formatter;
```
Comment (stadust) [inline src/serde/de/error.rs:3]: wrong import

File: src/serde/mod.rs [hunk 8]
```diff
@@ -1,6 +1,6 @@
-mod de;
-mod ser;
-mod thunk;
+pub(crate) mod de;
+pub(crate) mod ser;
+pub(crate) mod thunk;
```
Comment (stadust) [inline src/serde/mod.rs:3]: why?

File: src/serde/ser/error.rs [hunk 9]
```diff
@@ -1,7 +1,5 @@
-use std::{
-    fmt::{Display, Formatter},
-    io,
-};
+use serde::__private::Formatter;
```
Comment (stadust) [inline src/serde/ser/error.rs:1]: wrong import

File: src/serde/thunk.rs [hunk 10]
```diff
@@ -1,13 +1,7 @@
...
 use percent_encoding::{percent_decode_str, utf8_percent_encode, AsciiSet, CONTROLS};
-use serde::{ser::Error as _, Deserialize, Deserializer, Serialize, Serializer};
-use std::{
-    borrow::Cow,
-    fmt::{Display, Formatter},
-    num::ParseIntError,
-    str::Utf8Error,
-    string::FromUtf8Error,
-};
+use serde::{__private::Formatter, ser::Error as _, Deserialize, Deserializer, Serialize, Serializer};
```
Comment (stadust) [inline src/serde/thunk.rs:3]: wrong import

File: src/request/comment.rs [hunk 11]
```diff
@@ -220,4 +288,17 @@ mod tests {
...
+        if let Err(err) = env_logger::builder().is_test(true).try_init() {
+            // nothing to make the tests fail over
+            eprintln!("Error setting up env_logger: {:?}", err)
+        }
+
+        let request = CommentHistoryRequest::new(159782398)
+            .limit(2);
+
+        assert_eq!(super::super::to_string(request), "gameVersion=21&binaryVersion=33&secret=Wmfd2893gb7&total=0&page=0&mode=0&userID=159782398&count=2")
+    }
```
Comment (stadust) [inline src/request/comment.rs:303]: Tests go into the `tests` directory and are supposed to be roundtrip tests to ensure bijectivity of serialization and deserialization.

**Review by stadust — CHANGES_REQUESTED** [review]
Thanks for the PR!

I am not sure how many of these are actual problems, I have not worked on this code for over a year so I might remember some things wrongly. Use your own judgement when going through the comments!

File: src/model/comment/level.rs (hunk 2 above)
Comment (zRyder) [reply src/model/comment/level.rs:80]: The CommentHistory struct is removed now, I haven't made the commit yet but I will in a bit.

File: src/model/mod.rs (hunk 3 above)
Comment (zRyder) [reply src/model/mod.rs:75]: Leftover testing while I was trying to figure out the logic of the build script, I'll remove it in another commit.

File: src/model/user/profile.rs (hunk 4 above)
Comment (zRyder) [reply src/model/user/profile.rs:2]: Which import do we not need here? is it (De)Serialize?

File: src/response.rs (hunk 5 above)
Comment (zRyder) [reply src/response.rs:18]: __private::Formatter is needed, this was an issue from some time ago but this import randomly broke and suddenly needed the __private module before it for some reason, I can test if it is still needed.

File: src/response.rs (hunk 6 above)
Comment (zRyder) [reply src/response.rs:213]: Already unified in another commit that I will make shortly.

File: src/serde/de/error.rs (hunk 7 above)
Comment (zRyder) [reply src/serde/de/error.rs:3]: __private::Formatter is needed, this was an issue from some time ago but this import randomly broke and suddenly needed the __private module before it for some reason, I can test if it is still needed.

File: src/serde/mod.rs (hunk 8 above)
Comment (zRyder) [reply src/serde/mod.rs:3]: More leftovers from testing/researching

File: src/serde/ser/error.rs (hunk 9 above)
Comment (zRyder) [reply src/serde/ser/error.rs:1]: __private::Formatter is needed, this was an issue from some time ago but this import randomly broke and suddenly needed the __private module before it for some reason, I can test if it is still needed.

File: src/serde/thunk.rs (hunk 10 above)
Comment (zRyder) [reply src/serde/thunk.rs:3]: __private::Formatter is needed, this was an issue from some time ago but this import randomly broke and suddenly needed the __private module before it for some reason, I can test if it is still needed.

File: src/request/comment.rs (hunk 11 above)
Comment (zRyder) [reply src/request/comment.rs:303]: I thought there were other tests here? I'll move this one though and write a more fleshed out test case.

File: src/model/user/profile.rs (hunk 4 above)
Comment (zRyder) [reply src/model/user/profile.rs:2]: If this is about the __private::Formatter, serde complains when __private is not specified.

File: src/request/comment.rs (hunk 11 above)
Comment (zRyder) [reply src/request/comment.rs:303]: Fleshed out test case can be done later, as comment history relies on The LevelComment struct, it always fails the test due to the out of order indexes and the missing index 7 all together in the raw comment data. I'll rework this some other time.

Comment (zRyder) [reply src/request/comment.rs:303]: Also in some of the request modules, there are test modules within a given request module (The code linked above is an example) There are serialize test cases here, why?

File: src/model/user/profile.rs (hunk 4 above)
Comment (stadust) [reply src/model/user/profile.rs:2]: it should be `use std::fmt::Formatter` if I'm not mistaken, the `serde::__private::Formatter` is just a re-export of the standard library. For some reason IntelliJ prefers to import that one, no idea why.

**Pull request closed without merging**